Add AP_Servo_Telem

IamPete1 opened the pull request:
This adds a servo telem library. Logging only at this stage. There are some benefits, DroneCAN servo logging can be rate limited, this is currently not possible as each instance is logged when it arrives, so it is rate limited individual rather than as a block of each instance.

In the longer term scripting bindings will be added to get and provide this data. That will allow servo health checking and failsafes.

IamPete1 commented on libraries/AP_DroneCAN/AP_DroneCAN.h:
```diff
@@ -343,7 +346,11 @@ class AP_DroneCAN : public AP_CANDriver, public AP_ESC_Telem_Backend {
 
-#if AP_DRONECAN_VOLZ_FEEDBACK_ENABLED
+#if AP_DRONECAN_HIMARK_SERVO_SUPPORT && AP_SERVO_TELEM_ENABLED
+    Canard::ObjCallback<AP_DroneCAN, com_himark_servo_ServoInfo> himark_servo_ServoInfo_cb{this, &AP_DroneCAN::handle_himark_servoinfo};
```
himark servo was just broken, there was a function but the callback was never registered.

tridge replied:
this must have been a regression, it worked when it went in

IamPete1 replied:
In the change of DroneCAN library I think, I fixed the same thing for the volz message, https://github.com/ArduPilot/ardupilot/pull/27924 we should probably check all the rest....

IamPete1 replied:
I have checked the rest, were OK no others missing.

IamPete1 commented on libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:
```diff
@@ -361,46 +359,44 @@ void AP_PiccoloCAN::update()
-                );
+
+                const AP_Servo_Telem::TelemetryData telem_data {
+                    .command_position = servo.commandedPosition(),
+                    .measured_position = servo.position(),
```
Strictly these are not the correct units, should be degrees. I'm not familiar enough with piccolo CAN to know if we can do a conversion.

IamPete1 commented on libraries/AP_PiccoloCAN/AP_PiccoloCAN.h:
```diff
@@ -131,7 +131,6 @@ class AP_PiccoloCAN : public AP_CANDriver, public AP_ESC_Telem_Backend
     AP_Int16 _ecu_id;       //!< ECU Node ID
     AP_Int16 _ecu_hz;       //!< ECU update rate (Hz)
 
-    HAL_Semaphore _telem_sem;
```
This was only ever called in one place, so its not needed.

IamPete1 commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -389,37 +389,25 @@ void AP_Volz_Protocol::update()
-                telem.data[i].primary_current,
-                telem.data[i].secondary_current,
-                telem.data[i].primary_voltage,
-                telem.data[i].secondary_voltage,
```
We do loose logging of the secondary current and voltage, they could be added as extra fields in the future.

tridge commented:
@IamPete1 I have added simulation of DroneCAN actuator status and fixed some bugs in the PR here:
https://github.com/tridge/ardupilot/commits/servo_telem/

tridge commented on libraries/AP_DroneCAN/AP_DroneCAN.cpp:
```diff
@@ -1374,62 +1375,83 @@ void AP_DroneCAN::handle_traffic_report(const CanardRxTransfer& transfer, const
+        .force = msg.force,
+        .speed = msg.speed,
+        .duty_cycle = msg.power_rating_pct,
+        .valid_types = AP_Servo_Telem::TelemetryData::Types::COMMANDED_POSITION |
```
this should be MEASURED_POSITION

tridge commented on libraries/AP_Servo_Telem/AP_Servo_Telem.cpp:
```diff
@@ -0,0 +1,160 @@
+            force       : telemdata.present(TelemetryData::Types::FORCE)             ? telemdata.force                         : AP::logger().quiet_nanf(),
+            speed       : telemdata.present(TelemetryData::Types::SPEED)             ? telemdata.speed                         : AP::logger().quiet_nanf(),
+            power_pct   : telemdata.duty_cycle,
+            pos_cmd     : telemdata.present(TelemetryData::Types::MEASURED_POSITION) ? telemdata.command_position              : AP::logger().quiet_nanf(),
```
should be COMMANDED

tridge commented on libraries/AP_Servo_Telem/AP_Servo_Telem.cpp:
```diff
@@ -0,0 +1,160 @@
+#if HAL_LOGGING_ENABLED
+void AP_Servo_Telem::write_log()
+{
+    AP_Logger *logger = AP_Logger::get_singleton();
```
we should have an active_mask and return immediately for the most common case of no actuators active

IamPete1 commented:
> @IamPete1 I have added simulation of DroneCAN actuator status and fixed some bugs in the PR here: https://github.com/tridge/ardupilot/commits/servo_telem/

Thanks, I have pulled in your changes.

andyp1per commented on libraries/AP_Logger/README.md:
```diff
@@ -68,6 +68,7 @@ Please keep the names consistent with Tools/autotest/param_metadata/param.py:33
 | 's' | "s" | seconds|
 | 'q' | "rpm" | revolutions per minute|  Not an SI unit, but sometimes more intuitive than Hertz|
 | 'r' | "rad" | radians|
+| 't' | "N.m" | Newton meters | torque |
```
This should be "N-m" or "N m"

IamPete1 replied:
Yeah, I agree, it's just matched the existing ones.
![SmartSelect_20241125_142833_Chrome](https://github.com/user-attachments/assets/7ce7fac3-1f2b-4f51-9552-ece7b1f0aa40)

andyp1per reviewed (approved):
Very nice!

tridge commented:
@IamPete1 nice work, thanks!

Ryanf55 commented:
Unfortunately, this broke the DDS build.

``` 
libraries/AP_DDS/generated/ardupilot_msgs/msg/GlobalPosition.h:47:15: error: expected identifier before numeric constant
   47 | #define FORCE 512
      |               ^~~
../../libraries/AP_Servo_Telem/AP_Servo_Telem.h:45:13: note: in expansion of macro ‘FORCE’
   45 |             FORCE              = 1 << 2,
      |             ^~~~~
     
```

I'm not happy about the generator using global defines, but it's written for C... Is an easy fix to rename the enum in this class?
